pkg/controller/storagecluster/cephcluster.go

```diff
@@ -366,24 +367,32 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster) []rook.StorageClassDevi
 
 		for i := 0; i < replica; i++ {
 			placement := rook.Placement{}
+			preparePlacement := rook.Placement{}
 
 			if noPlacement {
 				in := getPlacement(sc, "osd")
 				(&in).DeepCopyInto(&placement)
 
+				if noPreparePlacement {
+					in := getPlacement(sc, "osd-prepare")
+					(&in).DeepCopyInto(&preparePlacement)
+				}
+
 				if len(topologyKeyValues) >= replica {
-					podAffinityTerms := placement.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution
-					podAffinityTerms[0].PodAffinityTerm.TopologyKey = topologyKey
-
-					topologyIndex := i % len(topologyKeyValues)
-					nodeZoneSelector := corev1.NodeSelectorRequirement{
-						Key:      topologyKey,
-						Operator: corev1.NodeSelectorOpIn,
-						Values:   []string{topologyKeyValues[topologyIndex]},
+					setTopologyForAffinity(&placement, i, topologyKeyValues, topologyKey)
```

Comment:
Since the value of `topologyIndex` would be the same for both function calls, I would rather see `topologyIndex := i % len(topologyKeyValues)` out here and just pass `topologyKeyValues[topologyIndex]` directly rather than the list and index number.